### PR TITLE
PyTorch 1.3.1

### PR DESCRIPTION
**Pull request by akartsky**
Added PyTorch 1.3.1 support.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docker/1.3.0/py3/Dockerfile.gpu [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	                              mkl==2019.4 \
+	                              mkl-include==2019.4 \
+	                              cython==0.29.12 \
+	                              typing==3.6.4 && \
+	 /opt/conda/bin/conda clean -ya
+ENV PATH /opt/conda/bin:$PATH
+
+ARG PYTORCH_VERSION=1.3.0
+ARG TORCHVISION_VERSION=0.4.1
+ARG MMS_VERSION=1.0.7
```
Comment (nskool) [inline docker/1.3.0/py3/Dockerfile.gpu]: Please update this to the latest 1.0.8 everywhere.

File: docker/1.3.0/py3/Dockerfile.gpu [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	                              scipy==1.3.0 \
+	                              ipython==7.7.0 \
+	                              mkl==2019.4 \
+	                              mkl-include==2019.4 \
+	                              cython==0.29.12 \
+	                              typing==3.6.4 && \
+	 /opt/conda/bin/conda clean -ya
+ENV PATH /opt/conda/bin:$PATH
+
+ARG PYTORCH_VERSION=1.3.0
```
Comment (nskool) [inline docker/1.3.0/py3/Dockerfile.gpu]: Please update to PyTorch 1.3.1 as discussed, thank you.

File: docker/1.3.1/py2/Dockerfile.cpu [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+LABEL maintainer="Amazon AI"
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    ca-certificates \
+    cmake \
+    curl \
+    git \
+    wget \
+    vim \
```
Comment (codewithsk) [inline docker/1.3.1/py2/Dockerfile.cpu]: Please sort apt-get libraries

File: docker/1.3.1/py2/Dockerfile.cpu [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    libxrender-dev \
+    build-essential \
+    zlib1g-dev \
+    libglib2.0-0 \
+    libgl1-mesa-glx \
+    openjdk-8-jdk-headless
+
+# See http://bugs.python.org/issue19846
+ENV LANG C.UTF-8
+
```
Comment (codewithsk) [inline docker/1.3.1/py2/Dockerfile.cpu]: Move all ARGs and ENVs to the top. ARGs before ENV. Open MP envs are an exception.

File: docker/1.3.1/py2/Dockerfile.cpu [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ENV LD_LIBRARY_PATH /opt/conda/lib/:$LD_LIBRARY_PATH
+
+# Add arguments to achieve the version, python and url
+ARG PYTHON_VERSION=2.7
+
+RUN curl -o ~/miniconda.sh -O  https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh  && \
```
Comment (codewithsk) [inline docker/1.3.1/py2/Dockerfile.cpu]: Use && in the beginning of the next line (applies to all multi-line RUNs in all docker files)

File: docker/1.3.1/py2/Dockerfile.cpu [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+     rm ~/miniconda.sh && \
+     /opt/conda/bin/conda update conda && \
+     /opt/conda/bin/conda install -y python=$PYTHON_VERSION \
+     numpy==1.16.4 \
+     scipy==1.2.1 \
+     ipython==5.8.0 \
+     mkl==2019.4 \
+     mkl-include==2019.4 \
+     cython==0.29.12 \
+     typing==3.7.4 && \
```
Comment (codewithsk) [inline docker/1.3.1/py2/Dockerfile.cpu]: Update all libraries to latest versions

File: docker/1.3.1/py2/Dockerfile.cpu [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+ARG PYTORCH_VERSION=1.3.1
+ARG TORCHVISION_VERSION=0.4.2
+ARG MMS_VERSION=1.0.8
+RUN conda install -c conda-forge awscli==1.16.210 opencv==4.0.1 && \
+    conda install -y scikit-learn==0.20.3 \
+                     pandas==0.24.2 \
+                     pillow==6.1.0 \
+                     h5py==2.9.0 \
+                     requests==2.22.0 && \
+    conda install pytorch==$PYTORCH_VERSION torchvision==$TORCHVISION_VERSION cpuonly -c pytorch && \
```
Comment (codewithsk) [inline docker/1.3.1/py2/Dockerfile.cpu]: Align according to best practices doc

File: docker/1.3.1/py2/Dockerfile.cpu [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+COPY docker/$PYTORCH_VERSION/py2/mms-entrypoint.py /usr/local/bin/dockerd-entrypoint.py
+COPY docker/$PYTORCH_VERSION/py2/config.properties /home/model-server
+COPY src/sagemaker_pytorch_serving_container/deep_learning_container.py /usr/local/bin/deep_learning_container.py
+
+RUN chmod +x /usr/local/bin/dockerd-entrypoint.py && \
+    chmod +x /usr/local/bin/deep_learning_container.py
+
+COPY dist/sagemaker_pytorch_serving_container-1.2-py2.py3-none-any.whl /sagemaker_pytorch_serving_container-1.2-py2.py3-none-any.whl
+RUN pip install --no-cache-dir /sagemaker_pytorch_serving_container-1.2-py2.py3-none-any.whl && \
```
Comment (codewithsk) [inline docker/1.3.1/py2/Dockerfile.cpu]: Move the install to new line

File: docker/1.3.1/py2/Dockerfile.gpu [hunk 9]
```diff
@@ -0,0 +1,101 @@
+FROM nvidia/cuda:10.0-cudnn7-devel-ubuntu16.04
```
Comment (codewithsk) [inline docker/1.3.1/py2/Dockerfile.gpu]: Please make same changes as CPU to GPU.

File: docker/1.3.1/py2/Dockerfile.gpu [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+# See http://bugs.python.org/issue19846
+ENV LANG C.UTF-8
+
+ENV LD_LIBRARY_PATH /opt/conda/lib/:$LD_LIBRARY_PATH
+
+# Add arguments to achieve the version, python and url
+ARG PYTHON_VERSION=2.7
+
+# Install OpenSSH, Allow OpenSSH to talk to containers without asking for confirmation
+RUN apt-get install -y --no-install-recommends openssh-client openssh-server && \
```
Comment (codewithsk) [inline docker/1.3.1/py2/Dockerfile.gpu]: Move each install to one line.

File: docker/1.3.1/py2/mms-entrypoint.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import shlex
+import subprocess
+import sys
+import os.path
+
+from sagemaker_pytorch_serving_container import serving
```
Comment (codewithsk) [inline docker/1.3.1/py2/mms-entrypoint.py]: Should this be moved inside the if condition? Keep this consistent with the mms-entrypoint for mxnet.

File: docker/1.3.1/py3/Dockerfile.cpu [hunk 12]
```diff
@@ -0,0 +1,81 @@
+FROM ubuntu:16.04
```
Comment (codewithsk) [inline docker/1.3.1/py3/Dockerfile.cpu]: Same as previous docker files.

File: docker/1.3.1/py3/Dockerfile.gpu [hunk 13]
```diff
@@ -0,0 +1,93 @@
+FROM nvidia/cuda:10.0-cudnn7-devel-ubuntu16.04
```
Comment (codewithsk) [inline docker/1.3.1/py3/Dockerfile.gpu]: Same as previous docker file.

File: docker/1.3.1/py2/Dockerfile.cpu (hunk 6 above)
Comment (saimidu) [reply docker/1.3.1/py2/Dockerfile.cpu]: Easiest way to do this is to unpin all versions, run the build, if it's successful, pin versions according to the ones that come up in a `pip freeze | grep <package>` within the container.

File: docker/1.3.0/py3/Dockerfile.gpu (hunk 1 above)
Comment (akartsky) [reply docker/1.3.0/py3/Dockerfile.gpu]: done

File: docker/1.3.0/py3/Dockerfile.gpu (hunk 2 above)
Comment (akartsky) [reply docker/1.3.0/py3/Dockerfile.gpu]: done

File: docker/1.3.1/py2/Dockerfile.cpu (hunk 3 above)
Comment (akartsky) [reply docker/1.3.1/py2/Dockerfile.cpu]: done

File: docker/1.3.1/py2/Dockerfile.cpu (hunk 4 above)
Comment (akartsky) [reply docker/1.3.1/py2/Dockerfile.cpu]: done

File: docker/1.3.1/py2/Dockerfile.cpu (hunk 5 above)
Comment (akartsky) [reply docker/1.3.1/py2/Dockerfile.cpu]: done

File: docker/1.3.1/py2/Dockerfile.cpu (hunk 7 above)
Comment (akartsky) [reply docker/1.3.1/py2/Dockerfile.cpu]: done

File: docker/1.3.1/py2/Dockerfile.cpu (hunk 8 above)
Comment (akartsky) [reply docker/1.3.1/py2/Dockerfile.cpu]: done

File: docker/1.3.1/py2/Dockerfile.gpu (hunk 9 above)
Comment (akartsky) [reply docker/1.3.1/py2/Dockerfile.gpu]: done

File: docker/1.3.1/py2/Dockerfile.gpu (hunk 10 above)
Comment (akartsky) [reply docker/1.3.1/py2/Dockerfile.gpu]: done

File: docker/1.3.1/py3/Dockerfile.gpu (hunk 13 above)
Comment (akartsky) [reply docker/1.3.1/py3/Dockerfile.gpu]: done

File: docker/1.3.1/py3/Dockerfile.gpu [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    conda-forge \
+    awscli==1.16.210 \
+    opencv==4.0.1 \
+ && conda install -y \
+    scikit-learn==0.21.2 \
+    pandas==0.25.0 \
+    pillow==5.4.1 \
+    h5py==2.9.0 \
+    requests==2.22.0 \
+ && conda install -c \
```
Comment (nskool) [inline docker/1.3.1/py3/Dockerfile.gpu]: You can combine all commands of '-c pytorch' together i.e. magma-cuda100 is also being fetched from pytorch channel.

**Pull request merged**